### ISSUE TITLE
Add a warning for casts between slices with items of different size?

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/rust#41324]
If making a mistake similar to the one I made in https://github.com/rust-lang/rfcs/issues/1973, one might write code like:

```rust
use std::os::raw::c_char;
// foo: &[c_char]
unsafe {
    &*(foo as *const [c_char] as *const [char])
}
```

The `as` cast happily copies the length of the slice without considering what it means, and the newly-created slice refers to a chunk as memory four times bigger than the initial slice. This probably causes a buffer overflow when the new slice is used later.

The compiler could have caught this. I propose that `as` should warn (a hard error would be a breaking change) when casting from `*const [T]` to `*const [U]` (similarly `*mut`) if `size_of::<T>() != size_of::<U>` and suggest using `std::slice::from_raw_parts` (`from_raw_parts_mut`) with some length computation instead. (And maybe recommend an alignment assertion.)

**Comment by nodakai** [comment]
Actually I can't quite imagine when we'll ever have a legitimate need for such an obscure construct like `*const [T]` (even more obscure than trait objects which now have [`raw::TraitObject`](https://doc.rust-lang.org/std/raw/struct.TraitObject.html).) Can we deprecate casts into *any* slice fat pointers, if not remove them from the language?

Why isn't it enough to have `&[T]` with `.len()` and `.as_ptr()` methods? Am I missing anything?

**Comment by nagisa** [comment]
Converting from `&[u8]` to `&[i8]` via pointer casts is still a very valid use case. doing the same via `len/as_ptr/from_raw_parts` is significantly more effort. Besides it is consistent with an expectation that value type of pointer only really matters when dereferencing.

**Comment by bluss** [comment]
@nodakai It's very useful that for any pointer to T, we have `*const T/*mut T` with the same representation. This allows us to implement custom ownership semantics in Rust, like for example implementing `Rc<T>` by using raw pointers. Yes, also raw fat pointers.

**Comment by eddyb** [comment]
@nodakai Not only are we not deprecating them, they're *more idiomatic* than `transmute`.
Sadly some of the features aren't old enough for many people to have grown used to.
I mean, you can probably still find people introducing dynamic overhead because they haven't gotten the gospel of `extern "C"` functions with generic type parameters (the ABI has been orthogonal for years now).

**Comment by Mark-Simulacrum** [comment]
Based on discussion here, I'm going to close this -- I believe this is best handled by clippy, so please file an issue there if you'd like this.
